bin/hbase

```diff
@@ -136,6 +136,21 @@ if [ -f "$HBASE_HOME/conf/hbase-env-$COMMAND.sh" ]; then
   . "$HBASE_HOME/conf/hbase-env-$COMMAND.sh"
 fi
 
+# establish a default value for HBASE_OPTS if it's not already set. For now,
+# all we set is the garbage collector.
+if [ -z "${HBASE_OPTS}" ] ; then
+  major_version_number="$(parse_java_major_version "$(read_java_version)")"
+  case "$major_version_number" in
+  8|9|10)
+    HBASE_OPTS="-XX:+UseConcMarkSweepGC"
+    ;;
+  11|*)
+    HBASE_OPTS="-XX:+UseG1GC"
```

Comment:
G1GC is the default, why aren't we just leaving this out?

Reply:
I wonder if we should move GC selection into the `GC_OPTS` family of variables.

Reply:
I guess that would mean when a newer JDK release changes the default collector we'll stick with the G1GC we currently know. That's consistent with how we handled CMS before. Makes sense.

> I wonder if we should move GC selection into the GC_OPTS family of variables.

That is a really good point I had not thought of. Probably?

Should we even be setting a collector for the CLIENT_GC_OPTS?

Reply:
The benefit of setting it in `HBASE_OPTS` is that it gets picked up by all of our launch contexts. Moving it out to the `*_GC_OPTS` variables is probably a larger refactor effort.

Also not discussed here is the API compatibility around these variables. Moving GC selection out of `HBASE_OPTS` is definitely a conceptual change. Even changing the behavior from always-appends-the collector-selection to only-sets-when-`HBASE_OPTS`-is-empty is also compatibility change.

Reply:
> Also not discussed here is the API compatibility around these variables. Moving GC selection out of HBASE_OPTS is definitely a conceptual change. Even changing the behavior from always-appends-the collector-selection to only-sets-when-HBASE_OPTS-is-empty is also compatibility change.

Those are all operational compatibility, imho, so we're fine as long as we don't put it into a maintenance release and we include a good note.

Reply:
> The benefit of setting it in HBASE_OPTS is that it gets picked up by all of our launch contexts. Moving it out to the *_GC_OPTS variables is probably a larger refactor effort.

What do you mean by launch contexts here? I thought the intention of i.e. SERVER_GC_OPTS was that it would get included for the launching of any of our services?

Reply:
I like the idea of just leaving it out.

GC_OPTS would be new? Lets not if we can avoid it?

Reply:
> GC_OPTS would be new? Lets not if we can avoid it?

I believe Nick is referring to `SERVER_GC_OPTS` and `CLIENT_GC_OPTS` which we already define in hbase-env.sh and use in the main hbase launching script.

Reply:
Yes, there are at least these two that are exposed in https://github.com/apache/hbase/blob/master/conf/hbase-env.sh#L47-L71

Perhaps I'm proposing adding a new `GC_OPTS` or `HBASE_GC_OPTS` or `COMMON_GC_OPTS` where this collector specification is placed, and onto which `SERVER_GC_OPTS` and `CLIENT_GC_OPTS` are built.

Reply:
> What do you mean by launch contexts here?

I mean any of the various classes that are eventually launched by `bin/hbase`. Some paths append server-specific settings, client-specific, shell-specific, and others append nothing. From what I can tell, all of these start with `HBASE_OPTS`.

Reply:
Yes that is true. I don't know how wise it is for us to be setting the collector on the shell or the client, but /shrug not something we need to answer here.

Reply:
Just suggesting we not introduce new variables for operator to figure if we can avoid it.

Reply:
Leaving this as is for now. Will add a release note that we're no longer prescriptive of the collector if the user specified `HBASE_OPTS`.